Modules/posixmodule.c

```diff
@@ -7657,24 +7658,36 @@ check_CreateSymbolicLink(void)
     return Py_CreateSymbolicLinkW != NULL;
 }
 
+static int _PPathCchRemoveFileSpec_Initialized = 0;
+typedef HRESULT(__stdcall *PPathCchRemoveFileSpec)(PWSTR pszPath, size_t cchPath);
+static PPathCchRemoveFileSpec _PPathCchRemoveFileSpec;
+
 /* Remove the last portion of the path - return 0 on success */
-static int
+static void
 _dirnameW(WCHAR *path)
 {
     WCHAR *ptr;
-    size_t length = wcsnlen_s(path, MAX_PATH);
-    if (length == MAX_PATH) {
-        return -1;
+    size_t length = wcslen(path);
+
+    if (!_PPathCchRemoveFileSpec_Initialized) {
+        HMODULE pathapi = LoadLibraryW(L"api-ms-win-core-path-l1-1-0.dll");
+        if (pathapi) {
+            _PPathCchRemoveFileSpec = (PPathCchRemoveFileSpec)GetProcAddress(pathapi, "PathCchRemoveFileSpec");
+        }
+        else {
+            _PPathCchRemoveFileSpec = NULL;
+        }
+        _PPathCchRemoveFileSpec_Initialized = 1;
     }
 
-    /* walk the path from the end until a backslash is encountered */
-    for(ptr = path + length; ptr != path; ptr--) {
-        if (*ptr == L'\\' || *ptr == L'/') {
-            break;
+    if (_PPathCchRemoveFileSpec) {
+        if (FAILED(_PPathCchRemoveFileSpec(path, length))) {
+            Py_FatalError("Error removing path element in _dirnameW");
```

Comment:
I'm not excited about adding a fatal error in the midst of a Python call, especially since we have an error code here that could be returned instead (inside `getpathp.c` is different, because we aren't initialized enough to raise an exception at that stage).

Setting an error here and returning non-zero should work fine. I _believe_ all calls to this are already holding the GIL, but that would be the only issue.

Reply:
Thanks - I wasn't aware of the difference there. I'll update the diff.